cosmic-panel-bin/src/notifications.rs

```diff
@@ -0,0 +1,179 @@
+        .and_then(|f| fcntl::fcntl(fd, fcntl::FcntlArg::F_SETFD(f)));
+    let mut daemon_socket = match result {
+        // CLOEXEC worked and we can startup with session IPC
+        Ok(_) => unsafe { UnixStream::from_raw_fd(fd) },
+        // CLOEXEC didn't work, something is wrong with the fd, just close it
+        Err(err) => {
+            let _ = unistd::close(fd);
+            return Err(err).with_context(|| "Failed to setup session socket");
+        }
+    };
+
+    daemon_socket
+        .set_nonblocking(true)
+        .expect("Couldn't set nonblocking");
+    // read remaining bytes from socket
+    {
+        let mut buf = [0u8; 128];
+        while let Ok(size) = daemon_socket.read(&mut buf) {
+            if size == 0 {
+                break;
+            }
+        }
+    }
+    let data = ron::ser::to_string(&PanelRequest::Init)?;
+    daemon_socket.write_all(format!("{}\n", data).as_bytes())?;
+    let daemon_socket: Arc<UnixStream> = Arc::new(daemon_socket);
+    let daemon_socket_clone = daemon_socket.clone();
+    info!("Inserting channel into event loop");
+
+    // insert channel for spaces to send pending nortification appplet to
+    let (tx, rx) = calloop::channel::sync_channel(10);
+
+    loop_handle
+        .insert_source(
+            rx,
+            move |msg: calloop::channel::Event<PendingAppletEvent>, _, state| match msg {
+                calloop::channel::Event::Msg(msg) => match msg {
+                    PendingAppletEvent::Add(id, stream) => {
+                        info!("Received pending notification applet for space {}", &id);
+                        if !state.space.notification_applet_spaces.contains(&id) {
+                            // state.space.pending_notification_applet_ids.push(msg);
+                            if let Err(err) = write_socket(&daemon_socket, state, (id, stream)) {
+                                error!(
+                                    "Failed to send pending notification applet to daemon: {}",
+                                    err
+                                );
+                            }
+                        }
+                    }
+                    PendingAppletEvent::Remove(id) => {
+                        info!("Removing pending notification applet for space {}", &id);
+                        state.space.notification_applet_spaces.remove(&id);
+                    }
+                },
+                calloop::channel::Event::Closed => {
+                    warn!("Notification channel closed");
+                }
+            },
+        )
+        .map_err(|err| {
+            anyhow::anyhow!(
+                "Failed to insert notification channel into event loop: {}",
+                err
+            )
+        })?;
+
+    info!("Inserting session socket into event loop");
+    // insert source for daemon socket
+    loop_handle
+        .insert_source(
+            Generic::new(daemon_socket_clone, Interest::BOTH, Mode::Edge),
```

Comment:
Presumably only `Interest::READ` is needed here if no writing is done.

I guess either `Edge` or `Level` should work here. Though you have to be careful with edge-triggering. It's easier to make mistakes.

Reply:
Ya that was leftover from when I was writing here